src/sys/socket/addr.rs

```diff
@@ -687,27 +595,6 @@ impl UnixAddr {
     }
 }
 
-impl PartialEq for UnixAddr {
-    fn eq(&self, other: &UnixAddr) -> bool {
-        self.sun_path() == other.sun_path()
```

Comment:
So, I was reading up on `sockaddr_un` following [this link](http://man7.org/linux/man-pages/man7/unix.7.html). I don't know if we need to worry about the 2nd condition, where there is no path, and it's trivial to compare these. However, I'm not certain if this makes sense to implement in `libc`, or if this implementation should really live in `nix`. I don't know how much `libc`  wants to have standards logic in their code.

However, looking at our code, it doesn't appear that we properly extract the `sun_path` for abstract sockets via `UnixAddr::sun_path()`.

We have some tests involving the `sun_path` in `test/sys/test_socket.rs`, however they don't test the hash or equality functions of this struct. I'm thinking before continuing a discussion, I should add some tests for that. Do you agree? I'm looking at it however and I don't immediately know how these tests should be written. So if you have any ideas, @asomers, I'd love to hear them. Otherwise I'll take a stab at these.

Reply:
Yeah, we should definitely be testing equality for abstract sockets.  I would suggest just manually creating some so we can get all of the edge cases.

Reply:
Maybe we should suggest that libc not implement PartialEq or Hash for sockaddr_un, because it's so weird.

Reply:
> Yeah, we should definitely be testing equality for abstract sockets. I would suggest just manually creating some so we can get all of the edge cases.

Turns out we have some tests that check the paths for regular and abstract sockets. I was thinking of extending them, but I'm not certain the right way to do it. I was going to create a `u8` array of 0xFF the same size of the struct, transmute it to a `UnixAddr` and then manually set the `sockaddr_un` and `usize` fields. This will have a bunch of custom logic to set those fields the way I think they should be set.

What I would prefer to do is to initialize a `UnixAddr`, then set the stack to contain `0xFF` for a large chunk of memory, and initialize an identical `UnixAddr`, and then compare them for equality and hashing. That should test it very effectively and remove all of the complexity of stuffing data in correctly by hand. Do you know of a way to do that with Rust?

> Maybe we should suggest that libc not implement PartialEq or Hash for sockaddr_un, because it's so weird.

I'm not going to bother with this. There's no promise made there that equality isn't just binary equality, but Nix definitely implies

Reply:
> > Yeah, we should definitely be testing equality for abstract sockets. I would suggest just manually creating some so we can get all of the edge cases.
> 
> Turns out we have some tests that check the paths for regular and abstract sockets. I was thinking of extending them, but I'm not certain the right way to do it. I was going to create a `u8` array of 0xFF the same size of the struct, transmute it to a `UnixAddr` and then manually set the `sockaddr_un` and `usize` fields. This will have a bunch of custom logic to set those fields the way I think they should be set.
> 
> What I would prefer to do is to initialize a `UnixAddr`, then set the stack to contain `0xFF` for a large chunk of memory, and initialize an identical `UnixAddr`, and then compare them for equality and hashing. That should test it very effectively and remove all of the complexity of stuffing data in correctly by hand. Do you know of a way to do that with Rust?

That sounds cumbersome.  Instead of modifying those tests, why not write new tests directly in src/sys/socket/addr.rs?  That way you'll be able to access the underlying libc::sockaddr_un directly.